client/tests/phpunit/Form/Admin/ReportChecklistTypeTest.php

```diff
@@ -0,0 +1,101 @@
+<?php declare(strict_types=1);
+
+namespace AppBundle\Form\Admin;
+
+
+use AppBundle\Entity\Report\Checklist;
+use AppBundle\Entity\Report\Report;
+use Symfony\Component\Form\Test\TypeTestCase;
+
+class ReportChecklistTypeTest extends TypeTestCase
+{
+
+    /**
+     * @dataProvider formValuesProvider
+     */
+    public function testSubmitValidData(?string $costValues, bool $deputyDetails)
+    {
+        $report = new Report();
+        $report->setAvailableSections(['profDeputyCosts']);
+
+        $formDataObject = new Checklist($report);
+
+        $options = [
+            'report' => $report
+        ];
+
+        $form = $this->factory->create(ReportChecklistType::class, $formDataObject, $options);
+
+        $expected = new Checklist($report);
+        $expected->setPaymentsMatchCostCertificate($costValues);
+        $expected->setProfCostsReasonableAndProportionate($costValues);
+        $expected->setHasDeputyOverchargedFromPreviousEstimates($costValues);
+        $expected->setContactDetailsUptoDate($deputyDetails);
+        $expected->setDeputyFullNameAccurateInCasrec($deputyDetails);
+
+        $formData = [
+            'paymentsMatchCostCertificate' => $costValues,
+            'profCostsReasonableAndProportionate' => $costValues,
+            'hasDeputyOverchargedFromPreviousEstimates' => $costValues,
+        ];
+
+        $form->submit($formData);
+
+        $this->assertTrue($form->isSynchronized());
+
+        $this->assertEquals($expected, $formDataObject);
+    }
+
+    public function formValuesProvider()
+    {
+        return [
+            'Yes' => ['yes', false],
+            'No' => ['no', false],
+            'Null' => [null, false]
+        ];
+    }
+
+    /**
+     * @dataProvider invalidFormValuesProvider
+     */
+    public function testSubmitInvalidData($costCertificates, $reasonableCosts, $overChargedCosts, bool $deputyDetails)
```

Comment:
I'm not sure that this test has asserted that a validation error occurs, or perhaps I'm missing it? It seems to have the same assertions as the test above.

Reply:
The assertion here was checking that the expected values for the properties passed invalid values was null (e.g. the form failed to set it). Not ideal and now that you've bought it up I think it would make more sense at the entity level.

Reply:
Ah I see. Perhaps we can leave this as a workaround until we decide a better way to test our constraint annotations 👍